### ISSUE TITLE
debug version of clangd is crazy huge

### ISSUE THREAD

**Issue by chouzz** [clangd/clangd#634]
Hello,

I build the debug version of clangd with cmake in win10 and Ubuntu 20.04, but the executable file of clangd is very large (1.8G), while the release version of clangd is only 55M, which is very abnormal.

I used GCC tool chain and clang tool-chain in Ubuntu 20.04, the result is the same. 
-  Debug version of clangd: 1.8G
-  RelWithDebInfo of clangd: 1.4G

and on windows 10,  the debug version of clangd build with vs2019 is only 155M.

my cmake command:
```bash
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra" -DLLVM_TARGETS_TO_BUILD="X86" -DCLANG_ENABLE_CLANGD=ON  -DLLVM_ENABLE_ZLIB=OFF -G Ninja ../llvm
ninja clangd
root@bae9c09e2504:~/llvm-project/build/bin# ll -h
total 1.5G
drwxr-xr-x 1 root root 4.0K Dec 23 10:45 ./
drwxr-xr-x 1 root root 4.0K Dec 23 10:35 ../
-rwxr-xr-x 1 root root  33M Dec 23 10:05 clang-tblgen*
-rwxr-xr-x 1 root root 1.4G Dec 23 10:45 clangd*
-rwxrwxrwx 1 root root 2.3K Dec 23 09:56 llvm-lit*
-rwxr-xr-x 1 root root  73M Dec 23 10:01 llvm-tblgen*

root@bae9c09e2504:~/llvm-project/build/lib# ll -h
total 5.5G
drwxr-xr-x 1 root root 4.0K Dec 23 10:40 ./
drwxr-xr-x 1 root root 4.0K Dec 23 10:35 ../
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Analysis/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 AsmParser/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 BinaryFormat/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Bitcode/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Bitstream/
drwxr-xr-x 1 root root 4.0K Dec 23 09:25 CMakeFiles/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 CodeGen/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 DWARFLinker/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 DebugInfo/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Demangle/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 ExecutionEngine/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Extensions/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 FileCheck/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Frontend/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 FuzzMutate/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 IR/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 IRReader/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 InterfaceStub/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 LTO/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 LineEditor/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Linker/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 MC/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 MCA/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Object/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 ObjectYAML/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Option/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Passes/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 ProfileData/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Remarks/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Support/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 TableGen/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Target/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Testing/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 TextAPI/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 ToolDrivers/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 Transforms/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 WindowsManifest/
drwxr-xr-x 1 root root 4.0K Dec 23 09:57 XRay/
drwxr-xr-x 1 root root 4.0K Dec 23 10:05 clang/
drwxr-xr-x 1 root root 4.0K Dec 23 09:56 cmake/
-rw-r--r-- 1 root root 4.3K Dec 23 09:57 cmake_install.cmake
-rw-r--r-- 1 root root 236M Dec 23 10:11 libLLVMAnalysis.a
-rw-r--r-- 1 root root 2.8M Dec 23 10:00 libLLVMBinaryFormat.a
-rw-r--r-- 1 root root  17M Dec 23 10:05 libLLVMBitReader.a
-rw-r--r-- 1 root root 820K Dec 23 10:00 libLLVMBitstreamReader.a
-rw-r--r-- 1 root root 132M Dec 23 10:05 libLLVMCore.a
-rw-r--r-- 1 root root  22M Dec 23 10:01 libLLVMDebugInfoCodeView.a
-rw-r--r-- 1 root root 2.2M Dec 23 10:00 libLLVMDebugInfoMSF.a
-rw-r--r-- 1 root root 3.1M Dec 23 09:58 libLLVMDemangle.a
-rw-r--r-- 1 root root 3.8M Dec 23 10:13 libLLVMFrontendOpenMP.a
-rw-r--r-- 1 root root  52M Dec 23 10:05 libLLVMMC.a
-rw-r--r-- 1 root root  20M Dec 23 10:06 libLLVMMCParser.a
-rw-r--r-- 1 root root  37M Dec 23 10:06 libLLVMObject.a
-rw-r--r-- 1 root root 2.2M Dec 23 10:00 libLLVMOption.a
-rw-r--r-- 1 root root  17M Dec 23 10:05 libLLVMProfileData.a
-rw-r--r-- 1 root root 6.2M Dec 23 10:01 libLLVMRemarks.a
-rw-r--r-- 1 root root  49M Dec 23 10:00 libLLVMSupport.a
-rw-r--r-- 1 root root  13M Dec 23 10:00 libLLVMTableGen.a
-rw-r--r-- 1 root root 6.3M Dec 23 10:00 libLLVMTableGenGlobalISel.a
-rw-r--r-- 1 root root 5.5M Dec 23 10:01 libLLVMTextAPI.a
-rw-r--r-- 1 root root 154M Dec 23 10:13 libLLVMTransformUtils.a
-rw-r--r-- 1 root root 123K Dec 23 10:05 libLLVMX86Info.a
-rw-r--r-- 1 root root 358M Dec 23 10:18 libclangAST.a
-rw-r--r-- 1 root root  26M Dec 23 10:18 libclangASTMatchers.a
-rw-r--r-- 1 root root  86M Dec 23 10:19 libclangAnalysis.a
-rw-r--r-- 1 root root  46M Dec 23 10:09 libclangBasic.a
-rw-r--r-- 1 root root 6.4M Dec 23 10:29 libclangCrossTU.a
-rw-r--r-- 1 root root 280M Dec 23 10:40 libclangDaemon.a
-rw-r--r-- 1 root root  97M Dec 23 10:15 libclangDriver.a
-rw-r--r-- 1 root root 5.2M Dec 23 10:18 libclangEdit.a
-rw-r--r-- 1 root root  32M Dec 23 10:21 libclangFormat.a
-rw-r--r-- 1 root root 122M Dec 23 10:28 libclangFrontend.a
-rw-r--r-- 1 root root  43M Dec 23 10:29 libclangIndex.a
-rw-r--r-- 1 root root  57M Dec 23 10:10 libclangLex.a
-rw-r--r-- 1 root root  87M Dec 23 10:27 libclangParse.a
-rw-r--r-- 1 root root 4.9M Dec 23 10:13 libclangRewrite.a
-rw-r--r-- 1 root root 464M Dec 23 10:25 libclangSema.a
-rw-r--r-- 1 root root  93M Dec 23 10:27 libclangSerialization.a
-rw-r--r-- 1 root root 855M Dec 23 10:35 libclangStaticAnalyzerCheckers.a
-rw-r--r-- 1 root root 218M Dec 23 10:32 libclangStaticAnalyzerCore.a
-rw-r--r-- 1 root root  34M Dec 23 10:35 libclangStaticAnalyzerFrontend.a
-rw-r--r-- 1 root root  30M Dec 23 10:35 libclangTidy.a
-rw-r--r-- 1 root root 104M Dec 23 10:37 libclangTidyAbseilModule.a
-rw-r--r-- 1 root root 8.6M Dec 23 10:36 libclangTidyAlteraModule.a
-rw-r--r-- 1 root root  75M Dec 23 10:36 libclangTidyAndroidModule.a
-rw-r--r-- 1 root root 8.9M Dec 23 10:36 libclangTidyBoostModule.a
-rw-r--r-- 1 root root 300M Dec 23 10:39 libclangTidyBugproneModule.a
-rw-r--r-- 1 root root  71M Dec 23 10:39 libclangTidyCERTModule.a
-rw-r--r-- 1 root root 113M Dec 23 10:38 libclangTidyCppCoreGuidelinesModule.a
-rw-r--r-- 1 root root  13M Dec 23 10:36 libclangTidyDarwinModule.a
-rw-r--r-- 1 root root  35M Dec 23 10:37 libclangTidyFuchsiaModule.a
-rw-r--r-- 1 root root  83M Dec 23 10:37 libclangTidyGoogleModule.a
-rw-r--r-- 1 root root  24M Dec 23 10:39 libclangTidyHICPPModule.a
-rw-r--r-- 1 root root  19M Dec 23 10:36 libclangTidyLLVMLibcModule.a
-rw-r--r-- 1 root root  31M Dec 23 10:37 libclangTidyLLVMModule.a
-rw-r--r-- 1 root root 8.5M Dec 23 10:36 libclangTidyLinuxKernelModule.a
-rw-r--r-- 1 root root  16M Dec 23 10:36 libclangTidyMPIModule.a
-rw-r--r-- 1 root root  86M Dec 23 10:36 libclangTidyMiscModule.a
-rw-r--r-- 1 root root 225M Dec 23 10:38 libclangTidyModernizeModule.a
-rw-r--r-- 1 root root  36M Dec 23 10:36 libclangTidyObjCModule.a
-rw-r--r-- 1 root root  14M Dec 23 10:36 libclangTidyOpenMPModule.a
-rw-r--r-- 1 root root  85M Dec 23 10:36 libclangTidyPerformanceModule.a
-rw-r--r-- 1 root root  15M Dec 23 10:36 libclangTidyPortabilityModule.a
-rw-r--r-- 1 root root 229M Dec 23 10:37 libclangTidyReadabilityModule.a
-rw-r--r-- 1 root root  66M Dec 23 10:35 libclangTidyUtils.a
-rw-r--r-- 1 root root 8.7M Dec 23 10:36 libclangTidyZirconModule.a
-rw-r--r-- 1 root root  30M Dec 23 10:29 libclangTooling.a
-rw-r--r-- 1 root root 3.3M Dec 23 10:15 libclangToolingCore.a
-rw-r--r-- 1 root root 2.3M Dec 23 10:15 libclangToolingInclusions.a
-rw-r--r-- 1 root root  61M Dec 23 10:30 libclangToolingRefactoring.a
-rw-r--r-- 1 root root  21M Dec 23 10:28 libclangToolingSyntax.a
-rw-r--r-- 1 root root  30M Dec 23 10:30 libclangTransformer.a
-rw-r--r-- 1 root root 1.5M Dec 23 10:32 libclangdNetwork.a
-rw-r--r-- 1 root root 171K Dec 23 10:40 libclangdRemoteIndex.a
-rw-r--r-- 1 root root 754K Dec 23 10:33 libclangdSqlDb.a
-rw-r--r-- 1 root root  12M Dec 23 10:33 libclangdSqlite.a
-rw-r--r-- 1 root root 4.1M Dec 23 10:32 libclangdSupport.a
```

**Comment by HighCommander4** [comment]
Perhaps on Windows shared linking is the default?

I build on Linux with `-DBUILD_SHARED_LIBS=ON` and my `clangd` binary is much smaller (40M), but given the sizes of the shared libraries it loads, it would not surprise me if in a statically linked build the binary were > 1 GB.

**Comment by sam-mccall** [comment]
This is definitely true. I think it's *probably* not something to solve at the clangd level, and it's not clear how important debug binary size is and why. So more context on what you're trying to do would be useful.

### What is the data?
The size is indeed almost all debug-info. On my machine debug clangd is 1.36G (config dependent) and bloaty says:
```
./bloaty ~/src/llvm-project/build-debug/bin/clangd
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  64.0%   891Mi   0.0%       0    .debug_info
   9.5%   131Mi   0.0%       0    .debug_str
   9.0%   124Mi   0.0%       0    .debug_line
   5.9%  81.5Mi  70.2%  81.5Mi    .text
   5.1%  70.3Mi   0.0%       0    .strtab
   2.9%  40.4Mi   0.0%       0    .debug_ranges
   1.2%  16.9Mi  14.5%  16.9Mi    .eh_frame
   0.9%  12.9Mi   0.0%       0    .symtab
   0.8%  11.1Mi   9.6%  11.1Mi    .rodata
   0.3%  4.22Mi   3.6%  4.22Mi    .eh_frame_hdr
   0.2%  3.18Mi   0.0%       0    .debug_abbrev
   0.2%  2.13Mi   1.8%  2.13Mi    .data.rel.ro
   0.1%   970Ki   0.0%       0    .debug_loc
   0.0%       0   0.1%  78.6Ki    .bss
   0.0%  69.1Ki   0.1%  69.0Ki    .rela.dyn
   0.0%  40.4Ki   0.0%  40.3Ki    .got
   0.0%  29.8Ki   0.0%  26.8Ki    [21 Others]
   0.0%  26.9Ki   0.0%  26.8Ki    .data
   0.0%  13.2Ki   0.0%  13.2Ki    .dynstr
   0.0%  9.60Ki   0.0%  9.54Ki    .dynsym
   0.0%  9.52Ki   0.0%       0    [Unmapped]
 100.0%  1.36Gi 100.0%   116Mi    TOTAL
```
So like 85% debug info as we expect.

### Why is it in the binary?

> and on windows 10, the debug version of clangd build with vs2019 is only 155M.

Three reasonable places to put debug info:
- in the executable file. This is default for gcc/clang.
- in a separate file corresponding to the executable. This is default for MSVC I believe (PDB file). There are ways to achieve this on linux but I don't think this is a standard configuration.
- in files corresponding to the object files. This is gcc/clang behavior with -gsplit-dwarf.

Each of these have strengths and weaknesses, and following the platform conventions by default seems reasonable.

### Why is the debug info larger on linux than other platforms?

(This is speculation, but I'd guess it's fairly likely the PDB file on windows is <1GB).
My understanding is dwarfv4 just isn't a terribly space-compact format for complex C++. dwarfv5 is supposed to be better, if you have a new toolchain you can probably enable that.

### Why is the debug info larger for clangd than other binaries?

85% seems reasonably typical for LLVM - `clang` and `opt` are about 83%.
I can imagine it's high compared to other codebases - LLVM style is very big on trying to achieve zero-cost abstractions with microoptimized data structures, encouraging lots of inlining, template metaprogramming etc. This will drive up the debug binary/info size as all this information that's supposed to be optimized away is preserved.

### Whose problem is it to solve?

TL;DR: ideally cmake/gcc/clang, but possibly llvm-project. 

This can be attacked in a few places:
 - clangd's own code - I think this is very unlikely to do anything - we don't have that much code, and it doesn't make heavy use of techniques likely to bloat debug info.
 - code we use from elsewhere in llvm-project (mostly clang and support+adt). This might have more effect, but it's slow going and will trade off debug info size vs opt binary performance, where llvm cares enormously about the latter.
 - cmake config for llvm-project. e.g. you could maybe make a Debug build split-dwarf by default where supported.
 - cmake itself - this is an even more natural place to improve defaults, but they also have reasons to be more conservative. [20256](https://gitlab.kitware.com/cmake/cmake/-/issues/20256) [21179](https://gitlab.kitware.com/cmake/cmake/-/issues/21179)
 - core toolchain (gcc/clang etc). Making dwarfv5/split-dwarf more widely available/default, some future dwarf6 etc.

### How serious is the problem?

Debug binaries aren't a product we expect to ship to anyone, so this is a problem to the extent it inconveniences developers.

It's something that's affected all of LLVM for quite a long time, which has a lot of experts on this sort of thing (not me!). So I suspect it's some combination of not-that-important, easy-to-work around, or hard to fix. But i've been wrong before!

**Comment by chouzz** [comment]
I add this option and rebuild clangd, the `clangd` binary is 39M now. 

Seems the default link of debug version is static. It's not about the system, it's about the compiler. I used to use `GCC` on windows, which also has this problem, but `MSVC` won't have this problem.

**Comment by decimad** [comment]
I too have 155M of clangd executable on windows ... plus the accompanying debug database of 1.4G

**Comment by njames93** [comment]
> I add `-DBUILD_SHARED_LIBS=ON` option and rebuild clangd, the `clangd` binary is 39M now.
> 
> Seems the default link of debug version is static. It's not about the system, it's about the compiler. I used to use `GCC` on windows, which also has this problem, but `MSVC` won't have this problem.
> 
> I just want to debug clangd, but too large binary clangd will slow down the loading of debugging information

`-DBUILD_SHARED_LIBS=ON` will result in a smaller binary, however it wont speed up loading the binary and debug information, more likely it will actually slow it down.

**Comment by HighCommander4** [comment]
> I just want to debug clangd, but too large binary clangd will slow down the loading of debugging information

One thing that has helped improve the loading time of debug information for me, was adding `set index-cache on` to my  `.gdbinit`. The first time loading debug info after the binary changes is still slow, but subsequent times (with the same binary) is much faster.

**Comment by chouzz** [comment]
Alright, thanks for clearing that up for me
